syntax = "proto3";

package p2r;

import "google/protobuf/empty.proto";

enum Cause {
    UNDEFINED = 0;
    SUCCESS = 1;
    TIMEOUT = 2;
    ERROR = 3;
    NO_RESOURCES = 4;
}

enum TerminateReason {
    UNKNOWN = 0;
    PLANNED = 1;
    UNPLANNED = 2;
}

message ConnectionId {
    message ProtocolVersion {
        int32 major = 1;
        int32 minor = 2;
    }
    ProtocolVersion protocol_version = 1;
    int32 fp_id = 2;
    int32 rm_id = 3;
}

message Response {
    Cause cause = 1;
    optional ConnectionId connection_id = 2;
}

message TerminateWarning {
    ConnectionId connection_id = 1;
    TerminateReason reason = 2;
    int32 warning_id = 3;
    int32 timeout = 4;
}

message TerminateCancel {
    ConnectionId connection_id = 1;
    int32 warning_id = 2;
}

message RestoreWarning {
    ConnectionId connection_id = 1;
    int32 timeout = 2;
}

message SpeedNotification {
    ConnectionId connection_id = 1;
    float speed = 2;
}

service P2R {
    rpc P2rTerminateWarning( TerminateWarning ) returns (Response) {}
    rpc P2rTerminateWarningCancel( TerminateCancel ) returns (google.protobuf.Empty) {}
    rpc P2rRestoreWarning( RestoreWarning ) returns (google.protobuf.Empty) {}
    rpc P2rSpeedLevelNotification(SpeedNotification) returns (google.protobuf.Empty) {}
}